### DISCUSSION TITLE
Grain boundary reconstruction and segment length

### DISCUSSION THREAD

**Discussion by GitHubUserWithoutAName** [mtex-toolbox/mtex#915] [Ask Anything] [answered]
Dear MTEX community,

this is not a question on how to perform a specific task with MTEX, but rather how MTEX achieves certain things "under the hood": Could anyone please give me a hint at how, in MTEX, ...

1. ... grain boundary segments are constructed from the information which pixels in an EBSD map belong to which grain and how these segments are represented in MTEX (maybe as splines?)?
2. ... grain boundary segment lengths (which can be accessed via `grainBoundary.segLength` are calculated? Does this involve integration along splines, or how is this achieved?

I am mostly interested in the mathematical approach / algorithms used for both those tasks (and related papers, if available), especially considering the problem of non-indexed data points and having to fill in missing data points.

I dug around the source code of MTEX as well as the documentation a little bit, but was not really able to figure this out.

Thanks a lot for any help anyone can offer, as always!

**Comment by ralfHielscher** [comment, marked as answer]
1. Have you had a look at the paper: [Grain detection from 2d and 3d EBSD data - Specification of the MTEX algorithm, F.Bachmann, R. Hielscher, H. Schaeben: Ultramicroscopy, 111, 1720-1733, 2011](https://www-user.tu-chemnitz.de/~rahi/paper/grains.pdf)
2. There are no splines at all. A segment is a straight line.

Currently we would like to improve our algorithm with respect to the following ideas:
1. Use the alphashape algorithm to detect the outer boundary better.
2. Implement an efficient algorithm to sort the boundary segments directly after generation. Currently, we do this only for small data sets just before plotting.
3. If 2. is in place we could subdivide boundary segments which, together with the smooth function, would allow for more smooth grain boundaries.

If you are interested in one of these aspects or have any other ideas for improving grain reconstruction I would be very happy to collaborate.

Ralf.

**Reply by Brian-Liu-rx** [reply]
Hi, could I understand in this way: A grain boundary segment is the edge length of one pixel unit (step size of EBSD scanning)?

**Reply by ralfHielscher** [reply]
No :) See the linked paper.

Ralf.

**Reply by GitHubUserWithoutAName** [reply]
Thanks a lot Ralf for your helpful reply and please excuse me for replying so late. The paper cleared things up nicely!

The reason I was asking: I am presently trying to come up with a way to extract certain grain parameters from gray-scale SEM images and was wondering if I could learn something from the approach MTEX is using. But it seems the MTEX approach cannot really be translated in that way, because the actual orientation information is needed.

Your suggestions certainly seem like a good idea!
